File: p4/orden_Superior/ordenSuperior.scala
// 4. Generar funciones de (orden superior) que permita 
//    realizar los solicitado en el punto 2. 

// Para compilar utilizamos :\orden_Superior> scalac ordenSuperior .\ordenSuperior.scala
// Para Ejecucion scala :\orden_Superior> scala ordenSuperior.scala
object ordenSuperior {

def suma(a:Int,b:Int):Int = a + b; 
def resta(a:Int,b:Int):Int=a-b
def multiplica(a:Int,b:Int):Int=a*b
def divide(a:Int,b:Int):Int=a/b
 //El parámetro de la función es la función
def fun(a:Int,b:Int,calcu:(Int,Int)=>Int)={
    calcu(a,b);
}


def main(args: Array[String]): Unit = {

     println( fun(8,2,suma));
     println( fun(8,2,resta));
     println( fun(8,2,multiplica));
     println( fun(8,2,divide));
}

}

File: p4/ordenSuperioH/ordenSuperior.hs
-- Calculadora Aritmética Con Haskell Aplicando (funciones de orden superior)
-- Para compilar utilizamos :\ordenSuperioH> ghc -o ordenSuperior .\ordenSuperior.hs
-- Para Ejecucion :\ordenSuperioH> .\ordenSuperior.exe

suma :: Integer ->(Integer -> Integer)   --declaración de función
suma  x y =  x + y                      --definición de función
resta :: Integer -> (Integer -> Integer)
resta x y =  x - y                     
multicar :: Integer ->(Integer -> Integer)  
multicar x y =  x * y   
dividir :: Integer -> (Integer -> Integer)
dividir x y =  x `div` y    


main = do 
    putStrLn "Calculadora Aritmetica"
    print((suma 8) 2)    --llamada a la función 
    print((resta 8) 2)   
    print((multicar 8) 2)    
    print((dividir 8 )2) 


